libraries/WiFi/src/WiFiGeneric.h

```diff
@@ -212,6 +218,7 @@ class WiFiGenericClass
 
   public:
     static int hostByName(const char *aHostname, IPAddress &aResult);
+    static int hostByName6(const char *aHostname, ip_addr_t& aResult);
```

Comment:
Why are you adding a separate interface just to call hostByName6?  Obviously no clients are using it (because you just added it). I suggest removing it as unnecessary.

The purpose of DNS is to be address-agnostic, a client should be able to just call hostByName() and get back the corresponding IPAddress (4 or 6).  (The priority order of addresses should be based on RFC 6724, usually IPv6 first if supported).

There should be no situations in which a separate call to hostByName6()  (or hostByName4()) needs to be called.

It is true that hosts may have multiple addresses (they may have multiple IPv6 and/or multiple IPv4), but this doesn't really address this. e.g. if the first one failed, maybe you want to check the second -- but that applies to multiple addresses of any type.

Even if you made it a more generic IP type parameter, if you want to pre-filter for IPv4 or IPv6, then I struggle to see when you would need to use this. i.e. how is the client code ever supposed to know when it only wants IPv6?

It really should not care. It should say "given me the IP address for this host".  If v6 is available, it will be returned, and if not then v4 will be.

Reply:
What about common use cases like `ping6 host` to explicitly check connectivity using IPv6? When I want to create a similar function, do you then need this or how can it be done? (asking out of curiosity)

Reply:
Okay, so what is the relevance of ping6? You might want to know if your device can connect to server "example.org"; i.e. you send a message and the server receives it. What does it matter how it gets there?  How is your app going to behave different?  What logic would you want to write where you app does, or doesn't, do something based on ping6?

i.e. there might be logic based on whether an address (6 or 4) is reachable or not, but I can't think of any scenario where there is a specific dependence on which protocol is used.

----

NOTE: This is different from saying if ping address1 fails, then try address2, because that is more generic than which IP version is being used, e.g. if a1 & a2 are both IPv4, then if a1 fails you want to use a2, or if a1 & a2 are both IPv6, then if a1 fails you want to use a2.   The scenario "if IPv6 fails use IPv4" is just a specific combination of the more generic check.  e.g. if an address has a1 IPv6, a2 IPv6, and a3 IPv4, and ping a1 fails, then you would want to try a2 (not a3).

Also, if there is no IPv6 network (you don't have an address to send from) and there is only IPv4, then even if there are DNS entries for both AAAA and A, then the address selection function should, according to RFC 6724, only give you the IPv4 address (as it is the only reachable address).

There are some cases where a server has multiple addresses and you want to check the up/down of all the addresses, but again IPv4/v6 is not relevant, e.g. they might have 2x IPv6 and 2x IPv4.


Maybe in a very rare diagnostic app you want to be able to report if any of the IPv6 are working or not (vs any of the IPv4), but again such an app would want to check all addresses, not just the first IPv6 and then jump to the first IPv4.

The other time 4 vs 6 might be relevant is if you are implementing happy eyeballs type algorithm, where it tries both and uses the first that responds. IoT devices don't usually have this, but I guess it could be a case (and even though you would probably want all IPv6 and all IPv4, not just the first of each).

Reply:
As a network administrator I do such tests and checks quite often. A lot of tools provide ways to explicitly select the IP Version.

Lets assume I want to build a small device that tracks and checks connectivity like the existing RIPE Probe. This device would be in need of a way to test each protocol while being connected dual stack.

Please note, I clearly do not state this is the default, but in my experience it's a thing you often need to debug things.

Reply:
Sure, diagnostics needs to cover both, and you probably want to report the stack as part of those diagnostics, but I don't know when you would want to write separate code paths, rather than one that simply returns them all.

e.g. say a server "example.org" has four configured addresses AAAA 2606:2800:220:1::1946, AAAA 2606:2400:1000:1::1946, A 93.184.216.34, A 193.184.216.35.

I can see the need to test all four network connections. Maybe you even want to filter on an attribute and only test the first two (the IPv6 ones).

At a code level, this only needs two method signatures:
  hostByName(name, &result) -- where you just want the first one of them, i.e. a basic connection
  hostsByName(name, results[], max_length), or similar -- get multiple.

You could always then filter results, if you wanted only the IPv6 addresses.

It would be a bit strange to build those filters into the signatures:
  hostByName(name, &result)
  hostByName6(name, &result)
  hostByNameSecondResult(name, &result)
  hostByName6SecondResult(name, &result)

And why write code for only two of those specific conditions (the first IP and the first IPv6).

Not to mention the logic to then handle if one of them is or isn't null (the host name may only have A, or only AAAA, or the network may be only IPv4 or only IPv6). That a lot of ifs and checking for nulls.

A simple hostsByName() handles all these cases, e.g. if the server has 3x AAAA and no A's, then it will return three records that you can loop through. If it only has a single A record, then only one item will be returned in results. The diagnostic code would be a lot simpler.

Not to mention that I think the proposed implementation of hostByName6() will return either IPv6 or IPv4 results, depending on what is available, i.e. if the server only has one A record, then it will return that IPv4 address.

Reply:
> If you are dual stack

I think that's the default for decades to come.
We will probably never completely get rid of IPv4.
I have dual-stack since 2010 here at home and only in the last few years I see people with new internet subscriptions to have some DS-lite or CG-NAT for IPv4. Some still don't even have IPv6.
I think DS-lite or CG-NAT for IPv4 is how it will be for the next decades.

Reply:
> At a code level, this only needs two method signatures:
> hostByName(name, &result) -- where you just want the first one of them, i.e. a basic connection
> hostsByName(name, results[], max_length), or similar -- get multiple.

That should be ok for what I have in mind. Thanks.

Reply:
Here in the Netherlands, we have a few providers offering true dual stack, but those are more like the "tech enthusiast" or "professional" subscriptions.
The large "home" providers offer CG-NAT + IPv6. This way you stil have IPv4 but more like a very powerful router performing NAT for you.
This way you can't do port forwarding for IPv4 and also keeping up a VPN to an IPv4 end point is next to impossible.

I don't think that's the same as NAT64 or XLAT464 as it simply is like sharing your old IPv4 NAT router with your neighbors (but then separated).
For the end user there isn't much difference between CG-NAT + IPv6 and true dual stack, apart from not having your own IPv4 public address and not being able to do any port forwarding.  (and the stability issues for services like VPN connections)